### PR TITLE
feat(logstash source): Add log namespace and schema def support

### PR DESCRIPTION
**Pull request by neuronull**
Adds log namespace and schema def support as described in https://github.com/vectordotdev/vector/pull/13720.

closes: https://github.com/vectordotdev/vector/issues/15026
epic: https://github.com/vectordotdev/vector/issues/13723

### CODE REVIEW

File: src/sources/logstash.rs [hunk 1]
```diff
@@ -6,14 +6,17 @@ use std::{
...
+use lookup::{event_path, metadata_path, owned_value_path, path, OwnedValuePath, PathPrefix};
 use smallvec::{smallvec, SmallVec};
 use snafu::{ResultExt, Snafu};
 use tokio_util::codec::Decoder;
-use vector_config::configurable_component;
-use vector_core::config::LogNamespace;
+use value::Kind;
+use vector_config::{configurable_component, NamedComponent};
+use vector_core::config::{LegacyKey, LogNamespace};
+use vector_core::schema::Definition;
```
Comment (neuronull) [inline src/sources/logstash.rs:19]: Yes, I will combine these.

File: src/sources/logstash.rs [hunk 2]
```diff
@@ -670,6 +751,63 @@ mod test {
...
+                    Kind::timestamp().or_undefined(),
+                )
+                .with_metadata_field(
+                    &owned_value_path!("vector", "ingest_timestamp"),
+                    Kind::timestamp(),
+                )
+                .with_metadata_field(
+                    &owned_value_path!(LogstashConfig::NAME, "host"),
+                    Kind::bytes(),
+                );
```
Comment (spencergilbert) [inline src/sources/logstash.rs:782]: Vaguely nit, but I find it useful to keep the fields grouped whenever possible. Less mental work to ensure all the expected fields per metadata group are present.

**Review by spencergilbert — APPROVED** [review]
LGTM - one opinion note, but not blocking.

File: src/sources/logstash.rs (hunk 2 above)
Comment (spencergilbert) [reply src/sources/logstash.rs]: <3

File: src/sources/logstash.rs [hunk 3]
```diff
@@ -670,6 +753,63 @@ mod test {
...
+            Kind::object(Collection::empty()),
+            [LogNamespace::Legacy],
+        )
+        .with_event_field(
+            &owned_value_path!("message"),
+            Kind::bytes(),
+            Some("message"),
+        )
+        .with_event_field(&owned_value_path!("source_type"), Kind::bytes(), None)
+        .with_event_field(&owned_value_path!("timestamp"), Kind::timestamp(), None)
```
Comment (neuronull) [inline src/sources/logstash.rs:808]: Flagging for myself to investigate, shouldn't the meaning be Some("timestamp") ?

**Review by StephenWakely — APPROVED** [review]
Looks good to me!

**Pull request merged**